File: packages/core/src/plugins/builtin/codex/event-mapper.ts
// packages/core/src/plugins/builtin/codex/event-mapper.ts
import { readFile } from 'node:fs/promises';
import { extname } from 'node:path';
import type { SessionSink } from '@qlan-ro/mainframe-types';
import type {
  ItemCompletedParams,
  ItemStartedParams,
  TurnCompletedParams,
  TurnStartedParams,
  ThreadStartedParams,
  TokenUsageUpdatedParams,
} from './types.js';
import type { PatchChangeKind, FileChangeItem, CollabAgentToolCallItem, TodoListItem } from './item-types.js';
import { parseUnifiedDiff } from '../../../messages/parse-unified-diff.js';
import { lookupAgentMetadata, describeAgent, agentTitle } from './thread-registry.js';
import { createChildLogger } from '../../../logger.js';

const log = createChildLogger('codex:events');

export interface CodexSessionState {
  threadId: string | null;
  currentTurnId: string | null;
  currentTurnPlan: { id: string; text: string } | null;
  lastUsage?: {
    input_tokens: number;
    output_tokens: number;
    cache_read_input_tokens?: number;
  };
  /** Tracks collabAgentToolCall item ids that already had a CollabAgent tool_use emitted. */
  openCollabCards?: Set<string>;
  /**
   * Maps child thread id → parent CollabAgent tool_use id.
   * Items arriving with a threadId in this map are tagged with parentToolUseId so the
   * desktop's groupTaskChildren() promotes the parent card to a TaskGroup.
   */
  collabChildThreads?: Map<string, string>;
  /**
   * Maps child thread id → spawn prompt, captured from `tool: "spawnAgent"` items
   * so the later `tool: "wait"` card can use the prompt as its description (since
   * `wait` items don't carry the prompt themselves).
   */
  spawnPrompts?: Map<string, string>;
}

export function handleNotification(method: string, params: unknown, sink: SessionSink, state: CodexSessionState): void {
  log.debug({ method }, 'codex notification: %s', method);

  switch (method) {
    case 'thread/started':
      return handleThreadStarted(params as ThreadStartedParams, sink, state);
    case 'turn/started':
      return handleTurnStarted(params as TurnStartedParams, state);
    case 'item/completed':
      return handleItemCompleted(params as ItemCompletedParams, sink, state);
    case 'item/plan/delta':
      return handlePlanDelta(params as { itemId: string; delta: string }, state);
    case 'turn/completed':
      return handleTurnCompleted(params as TurnCompletedParams, sink, state);
    case 'thread/tokenUsage/updated':
      return handleTokenUsage(params as TokenUsageUpdatedParams, sink, state);
    case 'thread/compacted':
      sink.onCompact();
      return;
    case 'item/started':
      return handleItemStarted(params as ItemStartedParams, sink, state);
    // TODO: future — map turn/diff/updated to file change tracking / context.updated
    // TODO: future — map turn/plan/updated to Plans panel structured plan state
    case 'turn/diff/updated':
    case 'turn/plan/updated':
    case 'thread/closed':
    case 'thread/status/changed':
    case 'item/agentMessage/delta':
    case 'item/commandExecution/outputDelta':
    case 'item/fileChange/outputDelta':
    case 'item/reasoning/summaryTextDelta':
    case 'item/reasoning/textDelta':
    case 'account/rateLimits/updated':
    case 'thread/name/updated':
      return; // silently ignore known-but-unhandled notifications
    default:
      if (method.startsWith('codex/event/')) return;
      log.debug({ method }, 'codex: unhandled notification');
  }
}

function handleThreadStarted(params: ThreadStartedParams, sink: SessionSink, state: CodexSessionState): void {
  state.threadId = params.thread.id;
  sink.onInit(params.thread.id);
}

function handleTurnStarted(params: TurnStartedParams, state: CodexSessionState): void {
  state.currentTurnPlan = null;
  state.currentTurnId = params.turn.id;
}

function handlePlanDelta(params: { itemId: string; delta: string }, state: CodexSessionState): void {
  const { itemId, delta } = params;
  const prev = state.currentTurnPlan;
  if (prev && prev.id === itemId) {
    state.currentTurnPlan = { id: itemId, text: prev.text + delta };
  } else {
    state.currentTurnPlan = { id: itemId, text: delta };
  }
}

function handleItemStarted(params: ItemStartedParams, sink: SessionSink, state: CodexSessionState): void {
  const { item } = params;
  if (item.type !== 'collabAgentToolCall') return;
  // `spawnAgent` is dispatch metadata only — stash its prompt for the later `wait` card.
  if (item.tool === 'spawnAgent') {
    if (!state.spawnPrompts) state.spawnPrompts = new Map();
    for (const childId of item.receiverThreadIds ?? []) {
      if (item.prompt) state.spawnPrompts.set(childId, item.prompt);
    }
    return;
  }
  // Only `wait` items render a card.
  emitCollabTaskGroupStart(item, sink, state);
}

function handleItemCompleted(params: ItemCompletedParams, sink: SessionSink, state: CodexSessionState): void {
  const { item, threadId } = params;

  // If this item came from a spawned sub-agent's thread, tag emitted blocks with
  // the parent CollabAgent's tool_use id so the renderer nests them.
  const parentToolUseId = state.collabChildThreads?.get(threadId);
  if (parentToolUseId) {
    sink = wrapSinkWithParentId(sink, parentToolUseId);
  }

  // Plan items arrive as a terminal `item/completed` with type === 'plan'.
  // They aren't part of the formal ThreadItem union (yet), so branch defensively
  // before the typed switch below.
  const itemAsUnknown = item as { id?: string; type?: string; text?: string };
  if (itemAsUnknown.type === 'plan' && typeof itemAsUnknown.text === 'string' && itemAsUnknown.id) {
    state.currentTurnPlan = { id: itemAsUnknown.id, text: itemAsUnknown.text };
    return;
  }

  switch (item.type) {
    case 'agentMessage':
      sink.onMessage([{ type: 'text', text: item.text }]);
      return;

    case 'reasoning':
      sink.onMessage([{ type: 'thinking', thinking: item.summary.join('\n') || item.content.join('\n') }]);
      return;

    case 'commandExecution':
      sink.onMessage([
        {
          type: 'tool_use',
          id: item.id,
          name: 'Bash',
          input: { command: item.command },
        },
      ]);
      sink.onToolResult([
        {
          type: 'tool_result',
          toolUseId: item.id,
          content: item.aggregatedOutput ?? '',
          isError: item.exitCode !== undefined && item.exitCode !== 0,
        },
      ]);
      return;

    case 'fileChange': {
      const isCompleted = item.status !== 'inProgress';
      const isError = item.status === 'failed' || item.status === 'declined';
      item.changes.forEach((change: FileChangeItem['changes'][number], index: number) => {
        const toolId = `${item.id}:${index}`;
        const isAdd = change.kind.type === 'add';
        const toolName = isAdd ? 'Write' : 'Edit';
        const input: Record<string, unknown> = isAdd
          ? { file_path: change.path, content: extractAddedContent(change.diff) }
          : {
              file_path: change.path,
              old_string: '',
              new_string: '',
              ...((change.kind as Extract<PatchChangeKind, { type: 'update' }>).move_path != null
                ? { move_path: (change.kind as Extract<PatchChangeKind, { type: 'update' }>).move_path }
                : {}),
            };
        sink.onMessage([{ type: 'tool_use', id: toolId, name: toolName, input }]);
        if (isCompleted) {
          const structuredPatch = parseUnifiedDiff(change.diff);
          sink.onToolResult([
            {
              type: 'tool_result',
              toolUseId: toolId,
              content: 'OK',
              isError,
              ...(structuredPatch.length ? { structuredPatch } : {}),
            },
          ]);
        }
      });
      return;
    }

    case 'imageGeneration': {
      // Codex emits the generated image inline as base64 in `result`. Prefer that;
      // fall back to reading `savedPath` from disk if the inline payload is missing.
      const prompt = item.revisedPrompt;
      const inline = item.result;
      const emit = (data: string, mediaType: string) => {
        const content: Parameters<SessionSink['onMessage']>[0] = [{ type: 'image', mediaType, data }];
        if (prompt) content.unshift({ type: 'text', text: prompt });
        sink.onMessage(content);
      };

      if (inline) {
        emit(inline, mediaTypeFromExtension(item.savedPath ?? '.png'));
        return;
      }

      const path = item.savedPath;
      if (!path) {
        log.warn({ id: item.id }, 'codex: imageGeneration missing both result and savedPath');
        return;
      }
      readFile(path)
        .then((bytes) => emit(bytes.toString('base64'), mediaTypeFromExtension(path)))
        .catch((err) => {
          log.warn({ err: String(err), path }, 'codex: failed to read generated image');
        });
      return;
    }

    case 'collabAgentToolCall': {
      // `spawnAgent` is dispatch metadata only — stash its prompt for the `wait` card.
      if (item.tool === 'spawnAgent') {
        if (!state.spawnPrompts) state.spawnPrompts = new Map();
        for (const childId of item.receiverThreadIds ?? []) {
          if (item.prompt) state.spawnPrompts.set(childId, item.prompt);
        }
        return;
      }
      // `wait` is the renderable card — open it (if started didn't) and close with the result.
      if (!state.openCollabCards?.has(item.id)) {
        emitCollabTaskGroupStart(item, sink, state);
      }
      const childId = item.receiverThreadIds?.[0];
      const subAgentMessage = childId ? (item.agentsStates?.[childId]?.message ?? null) : null;
      const isError = item.status === 'failed' || item.status === 'interrupted';
      sink.onToolResult([
        {
          type: 'tool_result',
          toolUseId: item.id,
          content: subAgentMessage ?? 'Sub-agent completed',
          isError,
        },
      ]);
      state.openCollabCards?.delete(item.id);
      // Stop routing further items from this spawn's child thread(s) and drop the prompt.
      for (const cid of item.receiverThreadIds ?? []) {
        state.collabChildThreads?.delete(cid);
        state.spawnPrompts?.delete(cid);
      }
      return;
    }

    case 'mcpToolCall': {
      const server = item.server ?? 'codex';
      const toolName = `mcp__${server}__${item.tool}`;
      sink.onMessage([
        {
          type: 'tool_use',
          id: item.id,
          name: toolName,
          input: item.arguments,
        },
      ]);
      sink.onToolResult([
        {
          type: 'tool_result',
          toolUseId: item.id,
          content: item.error ? (item.error.message ?? '') : JSON.stringify(item.result?.content ?? ''),
          isError: !!item.error,
        },
      ]);
      return;
    }

    case 'todoList': {
      const todos = normalizeTodoListItems(item);
      if (todos.length > 0) sink.onTodoUpdate(todos);
      return;
    }

    default:
      log.debug({ type: (item as { type: string }).type }, 'codex: unhandled item type');
  }
}

function normalizeTodoListItems(item: TodoListItem): import('@qlan-ro/mainframe-types').TodoItem[] {
  return item.items.map((t) => ({
    content: t.text,
    status: t.completed ? ('completed' as const) : ('pending' as const),
    activeForm: t.text,
  }));
}

function handleTurnCompleted(params: TurnCompletedParams, sink: SessionSink, state: CodexSessionState): void {
  state.currentTurnPlan = null;
  state.currentTurnId = null;
  const { turn } = params;
  const isError = turn.status === 'failed' || turn.status === 'interrupted';

  sink.onResult({
    total_cost_usd: 0,
    usage: state.lastUsage,
    subtype: isError ? 'error_during_execution' : undefined,
    is_error: isError,
  });
  state.lastUsage = undefined;
}

function handleTokenUsage(params: TokenUsageUpdatedParams, _sink: SessionSink, state: CodexSessionState): void {
  state.lastUsage = {
    input_tokens: params.usage.input_tokens,
    output_tokens: params.usage.output_tokens,
    cache_read_input_tokens: params.usage.cached_input_tokens,
  };
}

function mediaTypeFromExtension(path: string): string {
  switch (extname(path).toLowerCase()) {
    case '.png':
      return 'image/png';
    case '.jpg':
    case '.jpeg':
      return 'image/jpeg';
    case '.gif':
      return 'image/gif';
    case '.webp':
      return 'image/webp';
    default:
      return 'application/octet-stream';
  }
}

function emitCollabTaskGroupStart(item: CollabAgentToolCallItem, sink: SessionSink, state: CodexSessionState): void {
  if (!state.openCollabCards) state.openCollabCards = new Set();
  state.openCollabCards.add(item.id);
  // Register the spawned thread(s) so subsequent child items get tagged with
  // parentToolUseId and the desktop's groupTaskChildren() nests them under this card.
  if (!state.collabChildThreads) state.collabChildThreads = new Map();
  for (const childId of item.receiverThreadIds ?? []) {
    state.collabChildThreads.set(childId, item.id);
  }
  const childId = item.receiverThreadIds?.[0];
  // Same identity mapping as history.ts — subagent_type is the nickname, description
  // is the spawn prompt (more informative than the bare role).
  const meta = childId ? lookupAgentMetadata([childId]).get(childId) : undefined;
  const subagentType = agentTitle(meta) ?? describeAgent(meta) ?? 'Sub-agent';
  const prompt = (childId && state.spawnPrompts?.get(childId)) ?? item.prompt ?? '';
  const description = describeAgent(meta) ?? (prompt || subagentType);
  // Real subagent tool name. The desktop's groupTaskChildren() promotes this to a
  // TaskGroup card when child items arrive tagged with parentToolUseId.
  sink.onMessage([
    {
      type: 'tool_use',
      id: item.id,
      name: 'CollabAgent',
      input: { prompt, description, subagent_type: subagentType },
    },
  ]);
}

function wrapSinkWithParentId(sink: SessionSink, parentToolUseId: string): SessionSink {
  return {
    ...sink,
    onMessage: (blocks) => sink.onMessage(blocks.map((b) => ({ ...b, parentToolUseId }))),
    onToolResult: (results) => sink.onToolResult(results.map((r) => ({ ...r, parentToolUseId }))),
  };
}

/** Extract added lines from a unified diff for Write tool input.content. */
function extractAddedContent(diff: string): string {
  return diff
    .split('\n')
    .filter((line) => line.startsWith('+') && !line.startsWith('+++'))
    .map((line) => line.slice(1))
    .join('\n');
}

File: packages/core-rs/crates/mainframe-adapter-codex/src/event_mapper.rs
//! Ported from `packages/core/src/plugins/builtin/codex/event-mapper.ts`.
//!
//! Maps Codex app-server notifications onto `SessionSink` callbacks. Every
//! notification method is dispatched identically to the TS `handleNotification`;
//! unknown methods are logged at debug and skipped (never a hard error).

use std::sync::Arc;

use mainframe_adapter_api::SessionSink;
use serde_json::Value;

use crate::collab_card;
use crate::item_types::ThreadItem;
pub(crate) use crate::parent_id_sink::ParentIdSink;
use crate::quota_rate_limit::{
    has_recognized_window, normalize_rate_limit_snapshot, snapshot_has_window,
};
pub use crate::session_state::{CodexSessionState, CurrentTurnPlan, LastUsage};
use crate::thread_item_render::render_completed_item;
use crate::turn_lifecycle::{
    handle_plan_delta, handle_token_usage, handle_turn_completed, handle_turn_started,
};
use crate::types::{
    AccountRateLimitsUpdatedParams, ItemCompletedParams, ItemStartedParams, PlanDeltaParams,
    ThreadStartedParams, TokenUsageUpdatedParams, TurnCompletedParams, TurnStartedParams,
};

pub fn handle_notification(
    method: &str,
    params: &Value,
    sink: &Arc<dyn SessionSink>,
    state: &mut CodexSessionState,
) {
    tracing::debug!(module = "codex:events", method, "codex notification");

    match method {
        "thread/started" => {
            if let Ok(p) = serde_json::from_value::<ThreadStartedParams>(params.clone()) {
                handle_thread_started(p, sink, state);
            }
        }
        "turn/started" => {
            if let Ok(p) = serde_json::from_value::<TurnStartedParams>(params.clone()) {
                handle_turn_started(p, state);
            }
        }
        "item/completed" => {
            if let Ok(p) = serde_json::from_value::<ItemCompletedParams>(params.clone()) {
                handle_item_completed(p, sink, state);
            }
        }
        "item/plan/delta" => {
            if let Ok(p) = serde_json::from_value::<PlanDeltaParams>(params.clone()) {
                handle_plan_delta(p, state);
            }
        }
        "turn/completed" => {
            if let Ok(p) = serde_json::from_value::<TurnCompletedParams>(params.clone()) {
                handle_turn_completed(p, sink, state);
            }
        }
        "thread/tokenUsage/updated" => {
            if let Ok(p) = serde_json::from_value::<TokenUsageUpdatedParams>(params.clone()) {
                handle_token_usage(p, state);
            }
        }
        "thread/compacted" => crate::compaction::handle_compaction_completed(sink, state),
        "item/started" => {
            if let Ok(p) = serde_json::from_value::<ItemStartedParams>(params.clone()) {
                handle_item_started(p, sink, state);
            }
        }
        "account/rateLimits/updated" => {
            if let Ok(p) = serde_json::from_value::<AccountRateLimitsUpdatedParams>(params.clone())
            {
                handle_account_rate_limits_updated(p, sink);
            }
        }
        // Known-but-unhandled notifications — silently ignore.
        "turn/diff/updated"
        | "turn/plan/updated"
        | "thread/closed"
        | "thread/status/changed"
        | "item/agentMessage/delta"
        | "item/commandExecution/outputDelta"
        | "item/fileChange/outputDelta"
        | "item/reasoning/summaryTextDelta"
        | "item/reasoning/textDelta"
        | "thread/name/updated" => {}
        _ => {
            if method.starts_with("codex/event/") {
                return;
            }
            tracing::debug!(
                module = "codex:events",
                method,
                "codex: unhandled notification"
            );
        }
    }
}

fn handle_thread_started(
    params: ThreadStartedParams,
    sink: &Arc<dyn SessionSink>,
    state: &mut CodexSessionState,
) {
    state.thread_id = Some(params.thread.id.clone());
    sink.on_init(&params.thread.id);
}

fn handle_account_rate_limits_updated(
    params: AccountRateLimitsUpdatedParams,
    sink: &Arc<dyn SessionSink>,
) {
    let quota =
        normalize_rate_limit_snapshot(&params.rate_limits, chrono::Utc::now().timestamp_millis());
    // C2 (#268): a snapshot that recognizes zero windows must not ingest — it would
    // bump freshness with no data behind it. Warn only when slots were present but
    // unrecognized (a genuine format drift), staying quiet on a benign empty snapshot.
    if !has_recognized_window(&quota) {
        if snapshot_has_window(&params.rate_limits) {
            tracing::warn!(
                "codex rate limit: snapshot had windows but none were recognized; skipping ingest"
            );
        }
        return;
    }
    sink.on_provider_quota("codex", quota);
}

/// Which session a notification's `threadId` belongs to (task 17, todo #247):
/// the parent's own thread (or untagged, or pre-`thread/started`), a registered
/// child, or neither — an item from a thread nobody named must be dropped, not
/// leaked to the parent's transcript.
pub(crate) enum Owner {
    Parent,
    Child(String),
    Unknown,
}

pub(crate) fn resolve_owner(thread_id: Option<&str>, state: &CodexSessionState) -> Owner {
    if state.thread_id.is_none() {
        return Owner::Parent;
    }
    match thread_id {
        None => Owner::Parent,
        Some(tid) if state.thread_id.as_deref() == Some(tid) => Owner::Parent,
        Some(tid) if state.sub_agent_cards.contains_key(tid) => Owner::Child(tid.to_string()),
        Some(_) => Owner::Unknown,
    }
}

fn handle_item_started(
    params: ItemStartedParams,
    sink: &Arc<dyn SessionSink>,
    state: &mut CodexSessionState,
) {
    let owner = resolve_owner(params.thread_id.as_deref(), state);
    let Some(sink) = owner_sink(&owner, sink, state) else {
        return;
    };
    let sink = &sink;

    match serde_json::from_value::<ThreadItem>(params.item) {
        Ok(ThreadItem::ContextCompaction(_)) => {
            crate::compaction::handle_compaction_started(sink);
        }
        Ok(ThreadItem::CollabAgentToolCall(item)) => {
            collab_card::on_collab_tool_call(&item, collab_card::Phase::Started, sink, state);
        }
        // Every other item type renders from its terminal `item/completed` event.
        Ok(_) | Err(_) => {}
    }
}

fn handle_item_completed(
    params: ItemCompletedParams,
    sink: &Arc<dyn SessionSink>,
    state: &mut CodexSessionState,
) {
    let owner = resolve_owner(params.thread_id.as_deref(), state);
    let Some(sink) = owner_sink(&owner, sink, state) else {
        return;
    };
    let sink = &sink;

    // Plan deltas only ever describe the parent's own turn — a child's plan item
    // falls through to the unhandled-item-type debug log below instead.
    if matches!(owner, Owner::Parent)
        && let Some((id, text)) = plan_item_fields(&params.item)
    {
        state.current_turn_plan = Some(crate::session_state::CurrentTurnPlan { id, text });
        return;
    }

    match serde_json::from_value::<ThreadItem>(params.item.clone()) {
        Ok(item) => render_completed_item(item, params.thread_id.as_deref(), sink, state),
        Err(_) => {
            tracing::debug!(
                module = "codex:events",
                r#type = params
                    .item
                    .get("type")
                    .and_then(|v| v.as_str())
                    .unwrap_or(""),
                "codex: unhandled item type"
            );
        }
    }
}

/// Resolves an `Owner` into the sink an item/turn should render through:
/// `Unknown` drops the notification (returns `None`); `Child(t)` wraps `sink` in
/// a `ParentIdSink` tagged with that child's card id; `Parent` passes `sink`
/// through unchanged (cloning the `Arc`, not deep-copying the sink).
fn owner_sink(
    owner: &Owner,
    sink: &Arc<dyn SessionSink>,
    state: &CodexSessionState,
) -> Option<Arc<dyn SessionSink>> {
    match owner {
        Owner::Unknown => {
            tracing::debug!(
                module = "codex:events",
                "codex: dropping item from an unregistered thread"
            );
            None
        }
        Owner::Child(t) => {
            let card_id = state.sub_agent_cards.get(t)?.card_id.clone();
            Some(Arc::new(ParentIdSink::new(sink.clone(), card_id)))
        }
        Owner::Parent => Some(sink.clone()),
    }
}

/// Plan items arrive as a terminal `item/completed` with `type === "plan"` (not
/// part of the ThreadItem union) — checked before typed dispatch.
fn plan_item_fields(item: &Value) -> Option<(String, String)> {
    if item.get("type").and_then(|v| v.as_str()) != Some("plan") {
        return None;
    }
    let text = item.get("text").and_then(|v| v.as_str())?.to_string();
    let id = item.get("id").and_then(|v| v.as_str())?.to_string();
    Some((id, text))
}

// PORT STATUS: src/plugins/builtin/codex/event-mapper.ts (395 lines)
// confidence: medium
// todos: 0
// notes: handle_notification dispatches every method identically to the TS switch;
// notes: unknown methods debug-log once + skip. `wrapSinkWithParentId` becomes a
// notes: ParentIdSink newtype over Arc<dyn SessionSink> (delegates all callbacks;
// notes: transforms only on_message/on_tool_result). CodexSessionState uses
// notes: always-present empty HashSet/HashMap for the TS lazily-created Set/Map
// notes: fields. The imageGeneration savedPath disk-read fallback keeps the TS
// notes: async readFile via tokio::spawn + a hand-rolled base64 encoder (no base64
// notes: crate in the allowlist; inline path uses Codex's own base64 unchanged).
// notes: parse_unified_diff is the crate-local shim (see history.rs blocker note).
// notes: handle_turn_completed sends SessionResult.context_tokens = this turn's raw
// notes: input usage (None when no usage yet), resolving the TS sink's
// notes: `contextTokens === undefined → fall back to usage` path (event-handler.ts:366)
// notes: here because Option<i64> can't carry the undefined/null distinction downstream.
// notes: Tests in tests/event_mapper.rs (collab-agent-spawn + plan-item-capture +
// notes: turn-completed context/usage). `account/rateLimits/updated` moved out of
// notes: the silent-ignore arm into handle_account_rate_limits_updated, which
// notes: normalizes via quota_rate_limit and calls sink.on_provider_quota (no `?.`
// notes: needed — the trait's default no-op body covers sinks that don't override
// notes: it). Tested in tests/quota_notification.rs.
// notes: task 1 (todo #247) carved CodexSessionState/CurrentTurnPlan/LastUsage into
// notes: session_state.rs, ParentIdSink into parent_id_sink.rs, and the four turn/
// notes: usage/plan handlers into turn_lifecycle.rs; re-exported here so external
// notes: `event_mapper::X` call sites keep compiling.
